keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,423 @@
+
+    Is given an archive as input and provides a human-readable report as output
+    containing explanations for any identified issues.
+
+    ```bash
+    kubectl kudo diagnostics analyze cassandra_diagnostics.zip
+    ```
+
+### Operator developer experience
+
+To configure diagnostics globally, this KEP introduces an optional top-level
+`diagnostics` key in operator.yaml.
+
+#### Diagnostics collection
+
+The following diagnostics will be implicitly collected without any configuration
+from the operator developer:
+
+- Logs for deployed pods related to the KUDO Instance
+- YAML for created resources, including both spec and status, related
+  to the KUDO instance
+- Output of `kubectl describe` for all deployed resources related to the KUDO
+  Instance
+- Current plan status, if one exists, or the KUDO Instance
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
+- RBAC resources that are applicable to the KUDO controller manager
+- Current settings and version information for KUDO
+- Status of last preflight check run.
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle.resources`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - name: Zookeeper Configuration File
+        key: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: /opt/zookeeper/server.properties
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+            name: "{{ .InstanceName }}-zookeeper"
+      - name: DNS information for running pod
+        key: "dns-information"
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          objectRef:
```

Comment:
ya I like label selectors a lot as an option.

Reply:
(still, think reference-able objects is an even better long term solution, graph graph graph 📈 😉 )

Reply:
Added label selector as @zen-dog suggested, skipped matchExpression as its usage doesn't seem so clear at the moment
Removed templates and _name_ field, because
- we agree on disliking templates in operator.yaml
- I can be wrong here, but if objects can be named according to a certain template (pattern), they can as well be labelled according to the pattern
- in the worst case we'll just add them